File: java/0934-shortest-bridge.java
class Solution {
    public int shortestBridge(int[][] grid) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] vis = new boolean[grid.length][grid[0].length];
        
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    mark(grid, vis, q, i, j);
                    i = grid.length;
                    j = grid[0].length;
                }
            }
        }
        int count = 0;
        int[][] dir = {{0,1}, {1,0}, {-1,0}, {0,-1}};
        
        while (!q.isEmpty()) {
            int n = q.size();
            
            for (int k = 0; k < n; k++) {
                int[] curr = q.poll();
                
                for (int j = 0; j < 4; j++) {
                    int r = curr[0] + dir[j][0];
                    int c = curr[1] + dir[j][1];

                    if (r >= 0 && c >= 0 && r < grid.length && c < grid[0].length && !vis[r][c]) {
                        if (grid[r][c] == 1) return count;
                        vis[r][c] = true;
                        q.offer(new int[]{r,c});
                    }
                } 
            }
            count++;
        }
        return -1;
    }

    void mark(int[][] grid, boolean[][] vis , Queue<int[]> q, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == 0 || vis[i][j]) return;
        vis[i][j] = true;
        q.add(new int[]{i,j});
        mark(grid, vis, q, i + 1, j);
        mark(grid, vis, q, i - 1, j);
        mark(grid, vis, q, i, j + 1);
        mark(grid, vis, q, i, j - 1);

    } 
}

File: kotlin/0934-shortest-bridge.kt
class Solution {

    fun shortestBridge(grid: Array<IntArray>): Int {
        // q = [ (row,column,level) ]
        val queue: Queue<Triple<Int, Int, Int>> = LinkedList()
        val rowColumnDirections = arrayOf(
            intArrayOf(0, 1),
            intArrayOf(0, -1),
            intArrayOf(1, 0),
            intArrayOf(-1, 0)
        )

        fun dfs(startRow: Int, startColumn: Int) {
            if (
                startRow !in grid.indices ||
                startColumn !in grid[startRow].indices ||
                grid[startRow][startColumn] == 0 ||
                grid[startRow][startColumn] == VISITED_CELL
            ) return
            grid[startRow][startColumn] = VISITED_CELL
            queue.add(Triple(startRow, startColumn, 0))
            for ((rowDir, columDir) in rowColumnDirections) {
                dfs(startRow + rowDir, startColumn + columDir)
            }
        }
        // find one island and mark the cells with -1
        outer@ for (row in grid.indices) {
            inner@ for (column in grid[row].indices) {
                if (grid[row][column] != 1) continue
                dfs(row, column)
                break@outer
            }
        }
        while (queue.isNotEmpty()) {
            val (row, column, level) = queue.remove()
            for ((rowDir, colDir) in rowColumnDirections) {
                if (row + rowDir !in grid.indices) continue
                if (column + colDir !in grid[row + rowDir].indices) continue
                if (grid[row + rowDir][column + colDir] == VISITED_CELL) continue
                if (grid[row + rowDir][column + colDir] == 1) return level
                queue.add(Triple(row + rowDir, column + colDir, level + 1))
                grid[row + rowDir][column + colDir] = VISITED_CELL
            }
        }
        throw IllegalStateException("Two islands where expected to be in the grid.")
    }

    companion object {
        private const val VISITED_CELL = -1
    }
}

